lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -649,93 +659,108 @@ def _post(self, params: ApiParams, _) -> Response:
             API_METHOD.POST, ParamType.DATASET, params
         ).value
 
-        # Build an Elasticsearch instance to manage the bulk update
-        elastic = Elasticsearch(self.elastic_uri)
-        self.logger.info("Elasticsearch {} [{}]", elastic, VERSION)
+        if self.require_stable and dataset.state.mutating:
+            raise APIAbort(
+                HTTPStatus.CONFLICT, f"Dataset state {dataset.state.name} is mutating"
+            )
+
+        map = Metadata.getvalue(dataset=dataset, key=Metadata.INDEX_MAP)
 
-        # Internally report a summary of successes and Elasticsearch failure
-        # reasons: this will look something like
-        #
-        # {
-        #   "ok": {
-        #     "index1": 1,
-        #     "index2": 500,
-        #     ...
-        #   },
-        #   "elasticsearch failure reason 1": {
-        #     "index2": 5,
-        #     "index5": 10
-        #     ...
-        #   },
-        #   "elasticsearch failure reason 2": {
-        #     "index3": 2,
-        #     "index4": 15
-        #     ...
-        #   }
-        # }
         report = defaultdict(Counter)
         count = 0
         error_count = 0
 
-        # NOTE: because streaming_bulk is given a generator, and also
-        # returns a generator, we consume the entire sequence within the
-        # `try` block to catch failures.
-        try:
-            # Pass the bulk command generator to the helper
-            results = helpers.streaming_bulk(
-                elastic,
-                self.generate_actions(params.body, dataset),
-                raise_on_exception=False,
-                raise_on_error=False,
-            )
+        # If we don't have an Elasticsearch index-map, then the dataset isn't
+        # indexed and we skip the Elasticsearch actions.
+        if map:
```

Comment:
Instead of indenting all this code, could we consider short-circuiting the `if` statement to exit early when we don't have a index map?

Reply:
There is a bunch of common code which is executed whether we have a map or not, so trying to exit early is likely to make the code more complicated.

However, it probably would be helpful to refactor this function, breaking it up into a number of subroutines -- i.e., changing the indentation here would be less painful if the body of the `if` were just a subroutine call.  (167 lines is a bit long for a single function -- granted, half of those are comments, but, still, I count at least six "things" which the function is doing, and it's best to limit that number to no more than two or three.)

(FWIW, turning on the ignore whitespace option in "Diff View settings" pull-down makes this much easier to review.)

Reply:
We still need to do completion actions. And I'm not really enthusiastic about complicating the logic by pulling out small methods, nor convinced that'll actually make anything easier to follow or maintain. Then again, I'm still wearing my turkey PTO head and I'll give that some more consideration later ...

Reply:
Pulling out small methods _shouldn't_ alter the logic -- it just moves the existing code from one place to another  -- but the moved code doesn't have to be indented so far, and the reader doesn't have to consider it when reading the code in the caller.  The hard part is coming up with nice names for the new methods (and, sometimes, getting their interfaces set up properly).  The benefits, aside from making the original function easier to read, are that everything gets easier to (unit) test, because the new methods can be tested in isolation and the original function can be tested more simply and easily by mocking the new methods.

Reply:
On a macro scale, no; but the easiest part to separate, the `results` analysis, constructs the `summary` as well as maintaining the counters, and I didn't think the code was big enough or complicated enough to justify the overhead of bundling that in an `@dataclass`. It mostly amounts to moving the comments out of the way... but I suppose I'll give that a try.

Reply:
Wait, if at line 675 `map` evaluates to `False`, we still enter the `if` block and execute the code?

Isn't it the case that when `map` is false we just drop to line 759?

Or am I missing something about Python 3?

Reply:
> Isn't it the case that when `map` is false we just drop to line 759?

Yes, but that line is another conditional, and, if the map isn't required, then execution continues, produces a report, and returns normally.

Reply:
The point was that there's more to the method after the `elif`. But I get that the large comment block makes it perhaps less than trivial to follow. 😆

Reply:
Okay, I withdraw my line of inquiry.